File: source/modulo_components/test/cpp/test_component_interface.cpp
#include <gtest/gtest.h>

#include <chrono>

#include <modulo_core/EncodedState.hpp>
#include <modulo_core/translators/message_writers.hpp>
#include <modulo_utils/testutils/ServiceClient.hpp>

#include "test_modulo_components/component_public_interfaces.hpp"

#include <sensor_msgs/msg/image.hpp>

namespace modulo_components {

using namespace std::chrono_literals;

template<class NodeT>
class ComponentInterfaceTest : public ::testing::Test {
protected:
  static void SetUpTestSuite() { rclcpp::init(0, nullptr); }

  static void TearDownTestSuite() { rclcpp::shutdown(); }

  void SetUp() override {
    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
    this->node_ = std::make_shared<NodeT>("ComponentInterfacePublicInterface", rclcpp::NodeOptions());
    this->component_ = std::make_shared<ComponentInterfacePublicInterface>(this->node_);
    if (std::is_same<NodeT, rclcpp::Node>::value) {
      this->pub_type_ = modulo_core::communication::PublisherType::PUBLISHER;
    } else if (std::is_same<NodeT, rclcpp_lifecycle::LifecycleNode>::value) {
      this->pub_type_ = modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER;
    }
  }

  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
  std::shared_ptr<ComponentInterfacePublicInterface> component_;
  std::shared_ptr<NodeT> node_;
  modulo_core::communication::PublisherType pub_type_;
};

using NodeTypes = ::testing::Types<rclcpp::Node, rclcpp_lifecycle::LifecycleNode>;
TYPED_TEST_SUITE(ComponentInterfaceTest, NodeTypes);

TYPED_TEST(ComponentInterfaceTest, AddAssignment) {
  this->component_->add_assignment("an_assignment", state_representation::ParameterType::INT);
  const auto one = this->component_->assignments_.size();
  // adding an assignment with empty name should fail
  EXPECT_THROW(this->component_->add_assignment("", state_representation::ParameterType::INT), modulo_core::exceptions::AddAssignmentException);
  // adding an assignment with the same name should just overwrite
  this->component_->add_assignment("an_assignment", state_representation::ParameterType::INT);
  EXPECT_EQ(this->component_->assignments_.size(), one);
  // names should be cleaned up
  EXPECT_NO_THROW(this->component_->add_assignment("7cleEaGn_AaSssiGNgn#ment", state_representation::ParameterType::DOUBLE));
  EXPECT_NE(this->component_->assignments_.size(), one);
  // names without valid characters should fail
  EXPECT_THROW(this->component_->add_assignment("@@@@@@", state_representation::ParameterType::DOUBLE), modulo_core::exceptions::AddAssignmentException);

  auto assignment_iterator = this->component_->assignments_.find("an_assignment");
  auto another_assignment_iterator = this->component_->assignments_.find("clean_assignment");
  auto no_assignment_iterator = this->component_->assignments_.find("no_assignment");

  EXPECT_TRUE(assignment_iterator != this->component_->assignments_.end());
  EXPECT_TRUE(another_assignment_iterator != this->component_->assignments_.end());
  EXPECT_TRUE(no_assignment_iterator == this->component_->assignments_.end());
}

TYPED_TEST(ComponentInterfaceTest, TriggerAssignment) {
  this->component_->add_assignment("trigger_assignment_string", state_representation::ParameterType::STRING);
  this->component_->add_assignment("trigger_assignment_int", state_representation::ParameterType::INT);

  EXPECT_NO_THROW(this->component_->trigger_assignment("trigger_assignment_string", std::string("test")));
  EXPECT_NO_THROW(this->component_->trigger_assignment("trigger_assignment_int", 5));
  EXPECT_THROW(this->component_->trigger_assignment("trigger_assignment_string", 5), std::runtime_error);
  EXPECT_THROW(this->component_->trigger_assignment("trigger_assignment_int", std::string("test")), std::runtime_error);
}

TYPED_TEST(ComponentInterfaceTest, AddBoolPredicate) {
  this->component_->add_predicate("foo", true);
  auto predicate_iterator = this->component_->predicates_.find("foo");
  EXPECT_TRUE(predicate_iterator != this->component_->predicates_.end());
  EXPECT_TRUE(predicate_iterator->second.get_value());
}

TYPED_TEST(ComponentInterfaceTest, AddFunctionPredicate) {
  this->component_->add_predicate("bar", [&]() { return false; });
  auto predicate_iterator = this->component_->predicates_.find("bar");
  EXPECT_TRUE(predicate_iterator != this->component_->predicates_.end());
  EXPECT_FALSE(predicate_iterator->second.get_value());
}

TYPED_TEST(ComponentInterfaceTest, GetPredicateValue) {
  this->component_->add_predicate("foo", true);
  EXPECT_TRUE(this->component_->get_predicate("foo"));
  this->component_->add_predicate("bar", [&]() { return true; });
  EXPECT_TRUE(this->component_->get_predicate("bar"));
  // predicate does not exist, expect false
  EXPECT_FALSE(this->component_->get_predicate("test"));
  // error in callback function except false
  this->component_->add_predicate("error", [&]() -> bool { throw std::runtime_error("An error occurred"); });
  EXPECT_FALSE(this->component_->get_predicate("error"));
}

TYPED_TEST(ComponentInterfaceTest, SetPredicateValue) {
  this->component_->add_predicate("foo", true);
  this->component_->set_predicate("foo", false);
  EXPECT_FALSE(this->component_->get_predicate("foo"));
  // predicate does not exist so setting won't do anything
  this->component_->set_predicate("bar", true);
  EXPECT_FALSE(this->component_->get_predicate("bar"));
  this->component_->add_predicate("bar", true);
  this->component_->set_predicate("bar", [&]() { return false; });
  EXPECT_FALSE(this->component_->get_predicate("bar"));
}

TYPED_TEST(ComponentInterfaceTest, DeclareSignal) {
  this->component_->declare_input("input", "test");
  EXPECT_EQ(this->component_->template get_parameter_value<std::string>("input_topic"), "test");
  EXPECT_TRUE(this->component_->inputs_.find("input") == this->component_->inputs_.end());
  this->component_->declare_output("output", "test_again");
  EXPECT_EQ(this->component_->template get_parameter_value<std::string>("output_topic"), "test_again");
  EXPECT_TRUE(this->component_->outputs_.find("output") == this->component_->outputs_.end());
}

TYPED_TEST(ComponentInterfaceTest, AddRemoveInput) {
  auto data = std::make_shared<bool>(true);
  EXPECT_NO_THROW(this->component_->add_input("8_teEsTt_#1@3", data));
  EXPECT_FALSE(this->component_->inputs_.find("test_13") == this->component_->inputs_.end());
  EXPECT_EQ(this->component_->template get_parameter_value<std::string>("test_13_topic"), "~/test_13");

  EXPECT_NO_THROW(this->component_->template add_input<std_msgs::msg::Bool>(
      "9_tEestT_#1@5", [](const std::shared_ptr<std_msgs::msg::Bool>) {}, "/topic", true));
  EXPECT_FALSE(this->component_->inputs_.find("test_15") == this->component_->inputs_.end());
  EXPECT_EQ(this->component_->template get_parameter_value<std::string>("test_15_topic"), "/topic");

  // trying to add an input that already exists should fail
  this->component_->template add_input<std_msgs::msg::String>(
      "test_13", [](const std::shared_ptr<std_msgs::msg::String>) {});
  EXPECT_EQ(
      this->component_->inputs_.at("test_13")->get_message_pair()->get_type(),
      modulo_core::communication::MessageType::BOOL);

  // remove input
  this->component_->remove_input("test_13");
  EXPECT_TRUE(this->component_->inputs_.find("test_13") == this->component_->inputs_.end());

  EXPECT_NO_THROW(this->component_->add_input("sensor_msg_data", std::make_shared<sensor_msgs::msg::Image>()));
  EXPECT_FALSE(this->component_->inputs_.find("sensor_msg_data") == this->component_->inputs_.end());
}

TYPED_TEST(ComponentInterfaceTest, AddInputWithUserCallback) {
  auto state_data = std::make_shared<state_representation::CartesianState>("A");
  bool callback_triggered = false;
  auto user_callback = [&callback_triggered] {
    callback_triggered = true;
  };

  EXPECT_NO_THROW(this->component_->add_input("state", state_data, user_callback));
  EXPECT_FALSE(this->component_->inputs_.find("state") == this->component_->inputs_.end());
  auto callback =
      this->component_->inputs_.at("state")->template get_handler<modulo_core::EncodedState>()->get_callback();
  state_representation::CartesianState new_state("B");
  auto message = std::make_shared<modulo_core::EncodedState>();
  modulo_core::translators::write_message(*message, new_state, this->component_->node_clock_->get_clock()->now());
  EXPECT_STREQ(state_data->get_name().c_str(), "A");
  EXPECT_NO_THROW(callback(message));
  EXPECT_TRUE(callback_triggered);
  EXPECT_STREQ(state_data->get_name().c_str(), "B");

  // if the subscription callback fails (for example, the message is invalid), the user callback should be skipped
  message->data.clear();
  callback_triggered = false;
  EXPECT_NO_THROW(callback(message));
  EXPECT_FALSE(callback_triggered);
}

TYPED_TEST(ComponentInterfaceTest, AddService) {
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 0);
  EXPECT_EQ(static_cast<int>(this->component_->string_services_.size()), 0);

  auto empty_callback = []() -> ComponentServiceResponse {
    return {true, "test"};
  };
  EXPECT_NO_THROW(this->component_->add_service("empty", empty_callback));
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 1);
  EXPECT_NE(this->component_->empty_services_.find("empty"), this->component_->empty_services_.cend());

  auto string_callback = [](const std::string& payload) -> ComponentServiceResponse {
    return {true, payload};
  };
  EXPECT_NO_THROW(this->component_->add_service("string", string_callback));
  EXPECT_EQ(static_cast<int>(this->component_->string_services_.size()), 1);
  EXPECT_NE(this->component_->string_services_.find("string"), this->component_->string_services_.cend());

  // adding a service under an existing name should fail for either callback type but is exception safe
  EXPECT_NO_THROW(this->component_->add_service("empty", empty_callback));
  EXPECT_NO_THROW(this->component_->add_service("empty", string_callback));
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 1);
  EXPECT_EQ(static_cast<int>(this->component_->string_services_.size()), 1);

  EXPECT_NO_THROW(this->component_->add_service("string", empty_callback));
  EXPECT_NO_THROW(this->component_->add_service("string", string_callback));
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 1);
  EXPECT_EQ(static_cast<int>(this->component_->string_services_.size()), 1);

  // adding an empty service name should fail
  EXPECT_NO_THROW(this->component_->add_service("", empty_callback));
  EXPECT_NO_THROW(this->component_->add_service("", string_callback));
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 1);
  EXPECT_EQ(static_cast<int>(this->component_->string_services_.size()), 1);

  // adding a mangled service name should succeed under a sanitized name
  EXPECT_NO_THROW(this->component_->add_service("8_teEsTt_#1@3", empty_callback));
  EXPECT_EQ(static_cast<int>(this->component_->empty_services_.size()), 2);
  EXPECT_NE(this->component_->empty_services_.find("test_13"), this->component_->empty_services_.cend());
}

TYPED_TEST(ComponentInterfaceTest, CallEmptyService) {
  using namespace modulo_utils::testutils;
  auto empty_callback = []() -> ComponentServiceResponse {
    return {true, "test"};
  };
  EXPECT_NO_THROW(this->component_->add_service("empty", empty_callback));

  auto client = std::make_shared<ServiceClient<modulo_interfaces::srv::EmptyTrigger>>(
      rclcpp::NodeOptions(), "/" + std::string(this->component_->node_base_->get_name()) + "/empty");
  this->exec_->add_node(this->component_->node_base_);
  this->exec_->add_node(client);
  auto request = std::make_shared<modulo_interfaces::srv::EmptyTrigger::Request>();
  auto future = client->call_async(request);
  auto success = this->exec_->spin_until_future_complete(future, 1s);
  EXPECT_EQ(success, rclcpp::FutureReturnCode::SUCCESS);
  auto response = future.get();
  EXPECT_TRUE(response->success);
  EXPECT_EQ(response->message, "test");
}

TYPED_TEST(ComponentInterfaceTest, CallStringService) {
  using namespace modulo_utils::testutils;
  auto string_callback = [](const std::string& payload) -> ComponentServiceResponse {
    return {true, payload};
  };
  EXPECT_NO_THROW(this->component_->add_service("string", string_callback));

  auto client = std::make_shared<ServiceClient<modulo_interfaces::srv::StringTrigger>>(
      rclcpp::NodeOptions(), "/" + std::string(this->component_->node_base_->get_name()) + "/string");
  this->exec_->add_node(this->component_->node_base_);
  this->exec_->add_node(client);
  auto request = std::make_shared<modulo_interfaces::srv::StringTrigger::Request>();
  request->payload = "payload";
  auto future = client->call_async(request);
  auto success = this->exec_->spin_until_future_complete(future, 0.5s);
  EXPECT_EQ(success, rclcpp::FutureReturnCode::SUCCESS);
  auto response = future.get();
  EXPECT_TRUE(response->success);
  EXPECT_EQ(response->message, "payload");
}

TYPED_TEST(ComponentInterfaceTest, CreateOutput) {
  auto data = std::make_shared<bool>(true);
  EXPECT_NO_THROW(this->component_->create_output(this->pub_type_, "test", data, "/topic", true, true));
  EXPECT_FALSE(this->component_->outputs_.find("test") == this->component_->outputs_.end());
  EXPECT_EQ(this->component_->template get_parameter_value<std::string>("test_topic"), "/topic");
  EXPECT_TRUE(this->component_->periodic_outputs_.at("test"));

  auto pub_interface = this->component_->outputs_.at("test");
  if (std::is_same<TypeParam, rclcpp::Node>::value) {
    EXPECT_EQ(pub_interface->get_type(), modulo_core::communication::PublisherType::PUBLISHER);
  } else if (std::is_same<TypeParam, rclcpp::Node>::value) {
    EXPECT_EQ(pub_interface->get_type(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER);
  }
  EXPECT_EQ(pub_interface->get_message_pair()->get_type(), modulo_core::communication::MessageType::BOOL);
  EXPECT_THROW(pub_interface->publish(), modulo_core::exceptions::CoreException);

  EXPECT_NO_THROW(this->component_->create_output(this->pub_type_, "8_teEsTt_#1@3", data, "", true, false));
  EXPECT_FALSE(this->component_->periodic_outputs_.at("test_13"));
  EXPECT_NO_THROW(this->component_->publish_output("8_teEsTt_#1@3"));
  EXPECT_NO_THROW(this->component_->publish_output("test_13"));
  EXPECT_THROW(this->component_->publish_output(""), modulo_core::exceptions::CoreException);
}

TYPED_TEST(ComponentInterfaceTest, TF) {
  this->component_->add_tf_broadcaster();
  this->component_->add_static_tf_broadcaster();
  this->component_->add_tf_listener();
  auto send_tf = state_representation::CartesianPose::Random("test", "world");
  EXPECT_NO_THROW(this->component_->send_transform(send_tf));
  sleep(1);
  state_representation::CartesianPose lookup_tf;
  EXPECT_NO_THROW(lookup_tf = this->component_->lookup_transform("test", "world"));
  auto identity = send_tf * lookup_tf.inverse();
  EXPECT_FLOAT_EQ(identity.data().norm(), 1.);
  EXPECT_FLOAT_EQ(abs(identity.get_orientation().w()), 1.);

  sleep(1);
  EXPECT_THROW(
      lookup_tf = this->component_->lookup_transform("test", "world", 0.9),
      modulo_core::exceptions::LookupTransformException);

  auto send_static_tf = state_representation::CartesianPose::Random("static_test", "world");
  EXPECT_NO_THROW(this->component_->send_static_transform(send_static_tf));
  EXPECT_THROW(
      auto throw_tf = this->component_->lookup_transform("dummy", "world"),
      modulo_core::exceptions::LookupTransformException);

  EXPECT_NO_THROW(lookup_tf = this->component_->lookup_transform("static_test", "world"));
  identity = send_static_tf * lookup_tf.inverse();
  EXPECT_FLOAT_EQ(identity.data().norm(), 1.);
  EXPECT_FLOAT_EQ(abs(identity.get_orientation().w()), 1.);

  std::vector<state_representation::CartesianPose> send_tfs;
  send_tfs.reserve(3);
  for (std::size_t idx = 0; idx < 3; ++idx) {
    send_tfs.emplace_back(state_representation::CartesianPose::Random("test_" + std::to_string(idx), "world"));
  }
  EXPECT_NO_THROW(this->component_->send_transforms(send_tfs));
  for (const auto& tf : send_tfs) {
    lookup_tf = this->component_->lookup_transform(tf.get_name(), tf.get_reference_frame());
    identity = tf * lookup_tf.inverse();
    EXPECT_FLOAT_EQ(identity.data().norm(), 1.);
    EXPECT_FLOAT_EQ(abs(identity.get_orientation().w()), 1.);
  }

  std::vector<state_representation::CartesianPose> send_static_tfs;
  send_static_tfs.reserve(3);
  for (std::size_t idx = 0; idx < 3; ++idx) {
    send_static_tfs.emplace_back(
        state_representation::CartesianPose::Random("test_static_" + std::to_string(idx), "world"));
  }
  EXPECT_NO_THROW(this->component_->send_static_transforms(send_static_tfs));
  for (const auto& tf : send_static_tfs) {
    lookup_tf = this->component_->lookup_transform(tf.get_name(), tf.get_reference_frame());
    identity = tf * lookup_tf.inverse();
    EXPECT_FLOAT_EQ(identity.data().norm(), 1.);
    EXPECT_FLOAT_EQ(abs(identity.get_orientation().w()), 1.);
  }
}

TYPED_TEST(ComponentInterfaceTest, GetSetQoS) {
  auto qos = rclcpp::QoS(5);
  this->component_->set_qos(qos);
  EXPECT_EQ(qos, this->component_->get_qos());
}

TYPED_TEST(ComponentInterfaceTest, AddTrigger) {
  EXPECT_NO_THROW(this->component_->add_trigger("trigger"));
  ASSERT_FALSE(
      std::find(this->component_->triggers_.cbegin(), this->component_->triggers_.cend(), "trigger")
      == this->component_->triggers_.cend());
  EXPECT_FALSE(this->component_->get_predicate("trigger"));
  EXPECT_NO_THROW(this->component_->trigger("trigger"));
}
}// namespace modulo_components

File: source/modulo_components/test/python/test_component_interface.py
import time

import clproto
import numpy as np
import pytest
import rclpy
import state_representation as sr
from modulo_interfaces.srv import EmptyTrigger, StringTrigger
from modulo_components.component_interface import ComponentInterface
from modulo_core.exceptions import CoreError, LookupTransformError
from rclpy.qos import QoSProfile
from std_msgs.msg import Bool, String
from sensor_msgs.msg import JointState


def raise_(ex):
    raise ex


@pytest.fixture()
def component_interface(ros_context):
    yield ComponentInterface('component_interface')


def test_rate_parameter(ros_context):
    component_interface = ComponentInterface('component_interface')
    assert component_interface.get_parameter_value("rate") == 10.0
    assert component_interface.get_rate() == 10.0
    assert component_interface.get_period() == 0.1

    parameter_overrides = [rclpy.Parameter("rate", value=200.0)]
    component_interface = ComponentInterface('component_interface', parameter_overrides=parameter_overrides)
    assert component_interface.get_parameter_value("rate") == 200
    assert component_interface.get_rate() == 200.0
    assert component_interface.get_period() == 0.005


def test_add_bool_predicate(component_interface):
    component_interface.add_predicate('foo', True)
    assert 'foo' in component_interface._ComponentInterface__predicates.keys()
    assert component_interface._ComponentInterface__predicates['foo'].get_value()


def test_add_function_predicate(component_interface):
    component_interface.add_predicate('foo', lambda: False)
    assert 'foo' in component_interface._ComponentInterface__predicates.keys()
    assert not component_interface._ComponentInterface__predicates['foo'].get_value()


def test_get_predicate(component_interface):
    component_interface.add_predicate('foo', True)
    assert component_interface.get_predicate('foo')
    component_interface.add_predicate('bar', lambda: True)
    assert component_interface.get_predicate('bar')
    # predicate does not exist, expect false
    assert not component_interface.get_predicate('test')
    # error in callback function except false
    component_interface.add_predicate('error', lambda: raise_(RuntimeError("An error occurred")))
    assert not component_interface.get_predicate('error')


def test_set_predicate(component_interface):
    component_interface.add_predicate('foo', True)
    component_interface.set_predicate('foo', False)
    assert not component_interface.get_predicate('foo')
    # predicate does not exist so setting won't do anything
    component_interface.set_predicate('bar', True)
    assert not component_interface.get_predicate('bar')
    component_interface.add_predicate('bar', True)
    component_interface.set_predicate('bar', lambda: False)
    assert not component_interface.get_predicate('bar')


def test_declare_signal(component_interface):
    component_interface.declare_input("input", "test")
    assert component_interface.get_parameter_value("input_topic") == "test"
    assert "input" not in component_interface._ComponentInterface__inputs.keys()
    component_interface.declare_output("output", "test_again")
    assert component_interface.get_parameter_value("output_topic") == "test_again"
    assert "test_again" not in component_interface._ComponentInterface__outputs.keys()


def test_add_remove_input(component_interface):
    component_interface.add_input("8_teEsTt_#1@3", "test", Bool)
    assert "test_13" in component_interface._ComponentInterface__inputs.keys()
    assert component_interface.get_parameter_value("test_13_topic") == "~/test_13"

    component_interface.add_input("9_tEestT_#1@5", "test", Bool, default_topic="/topic")
    assert "test_15" in component_interface._ComponentInterface__inputs.keys()
    assert component_interface.get_parameter_value("test_15_topic") == "/topic"

    component_interface.add_input("test_13", "test", String)
    assert component_interface._ComponentInterface__inputs["test_13"].msg_type == Bool

    component_interface.remove_input("test_13")
    assert "test_13" not in component_interface._ComponentInterface__inputs.keys()


def test_add_service(component_interface, ros_exec, make_service_client):
    def empty_callback():
        return {"success": True, "message": "test"}

    component_interface.add_service("empty", empty_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 1
    assert "empty" in component_interface._ComponentInterface__services_dict.keys()

    def string_callback(payload: str):
        return {"success": True, "message": payload}

    component_interface.add_service("string", string_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 2
    assert "string" in component_interface._ComponentInterface__services_dict.keys()

    # adding a service under an existing name should fail for either callback type, but is exception safe
    component_interface.add_service("empty", empty_callback)
    component_interface.add_service("empty", string_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 2

    component_interface.add_service("string", empty_callback)
    component_interface.add_service("string", string_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 2

    # adding an empty service name should fail
    component_interface.add_service("", empty_callback)
    component_interface.add_service("", string_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 2

    # adding a mangled service name should succeed under a sanitized name
    component_interface.add_service("8_teEsTt_#1@3", empty_callback)
    assert len(component_interface._ComponentInterface__services_dict) == 3
    assert "test_13" in component_interface._ComponentInterface__services_dict.keys()

    client = make_service_client(
        {"/component_interface/empty": EmptyTrigger, "/component_interface/string": StringTrigger})
    ros_exec.add_node(component_interface)
    ros_exec.add_node(client)
    future = client.call_async("/component_interface/empty", EmptyTrigger.Request())
    ros_exec.spin_until_future_complete(future, timeout_sec=0.5)
    assert future.done() and future.result().success
    assert future.result().message == "test"

    future = client.call_async("/component_interface/string", StringTrigger.Request(payload="payload"))
    ros_exec.spin_until_future_complete(future, timeout_sec=0.5)
    assert future.done() and future.result().success
    assert future.result().message == "payload"


def test_create_output(component_interface):
    component_interface._ComponentInterface__create_output(
        "test", "test", Bool, clproto.MessageType.UNKNOWN_MESSAGE, "/topic", True, True)
    assert "test" in component_interface._ComponentInterface__outputs.keys()
    assert component_interface.get_parameter_value("test_topic") == "/topic"
    assert component_interface._ComponentInterface__outputs["test"]["message_type"] == Bool
    assert component_interface._ComponentInterface__periodic_outputs["test"]

    component_interface._ComponentInterface__create_output(
        "8_teEsTt_#1@3", "test", Bool, clproto.MessageType.UNKNOWN_MESSAGE, "", True, False)
    assert not component_interface._ComponentInterface__periodic_outputs["test_13"]
    component_interface.publish_output("8_teEsTt_#1@3")
    component_interface.publish_output("test_13")
    with pytest.raises(CoreError):
        component_interface.publish_output("")

    component_interface._ComponentInterface__create_output("test_custom", "test", JointState,
                                                           clproto.MessageType.UNKNOWN_MESSAGE, "/topic", True, True)
    assert "test_custom" in component_interface._ComponentInterface__outputs.keys()
    assert component_interface.get_parameter_value("test_custom_topic") == "/topic"
    assert component_interface._ComponentInterface__outputs["test_custom"]["message_type"] == JointState
    data = JointState()
    data.name = ["joint_1", "joint_2"]
    msg = JointState()
    component_interface._ComponentInterface__outputs["test_custom"]["translator"](msg, data)
    assert msg.name == data.name
    assert component_interface._ComponentInterface__periodic_outputs["test_custom"]


def test_tf(component_interface):
    component_interface.add_tf_broadcaster()
    component_interface.add_static_tf_broadcaster()
    component_interface.add_tf_listener()
    send_tf = sr.CartesianPose().Random("test", "world")
    send_static_tf = sr.CartesianPose().Random("static_test", "world")
    component_interface.send_transform(send_tf)
    component_interface.send_static_transform(send_static_tf)
    for i in range(10):
        rclpy.spin_once(component_interface)
    with pytest.raises(LookupTransformError):
        component_interface.lookup_transform("dummy", "world")
    lookup_tf = component_interface.lookup_transform("test", "world")
    identity = send_tf * lookup_tf.inverse()
    assert np.linalg.norm(identity.data()) - 1 < 1e-3
    assert abs(identity.get_orientation().w) - 1 < 1e-3

    time.sleep(1.0)
    with pytest.raises(LookupTransformError):
        component_interface.lookup_transform("test", "world", validity_period=0.9)

    lookup_tf = component_interface.lookup_transform("static_test", "world")
    identity = send_static_tf * lookup_tf.inverse()
    assert np.linalg.norm(identity.data()) - 1 < 1e-3
    assert abs(identity.get_orientation().w) - 1 < 1e-3

    send_tfs = []
    for idx in range(3):
        send_tfs.append(sr.CartesianPose().Random("test_" + str(idx), "world"))
    component_interface.send_transforms(send_tfs)
    for i in range(10):
        rclpy.spin_once(component_interface)
    for tf in send_tfs:
        lookup_tf = component_interface.lookup_transform(tf.get_name(), tf.get_reference_frame())
        identity = tf * lookup_tf.inverse()
        assert np.linalg.norm(identity.data()) - 1 < 1e-3
        assert abs(identity.get_orientation().w) - 1 < 1e-3

    send_static_tfs = []
    for idx in range(3):
        send_static_tfs.append(sr.CartesianPose().Random("test_static_" + str(idx), "world"))
    component_interface.send_static_transforms(send_static_tfs)
    for i in range(10):
        rclpy.spin_once(component_interface)
    for tf in send_static_tfs:
        lookup_tf = component_interface.lookup_transform(tf.get_name(), tf.get_reference_frame())
        identity = tf * lookup_tf.inverse()
        assert np.linalg.norm(identity.data()) - 1 < 1e-3
        assert abs(identity.get_orientation().w) - 1 < 1e-3


def test_get_set_qos(component_interface):
    qos = QoSProfile(depth=5)
    component_interface.set_qos(qos)
    assert qos == component_interface.get_qos()


def test_add_trigger(component_interface):
    component_interface.add_trigger("trigger")
    assert "trigger" in component_interface._ComponentInterface__triggers
    assert not component_interface.get_predicate("trigger")
    component_interface.trigger("trigger")
